File: backend/app/services/layout.py
"""The grid a board is arranged on.

The number of columns is a setting of the board, not of the program. A board
without the setting has twelve, which is what every board arranged before
HexDeck was drawn on; new boards get twenty-four, half a column being the
difference between a card that fits and one that does not on a wide screen.

⚠️ Adapters declare their sizes in twelfths and are not touched: the 140 of
them are upstream's, and every one changed is a merge conflict. The scaling
happens here, at the two places a size meets a layout.
"""

from __future__ import annotations

from typing import Any

ALLOWED_COLUMNS: frozenset[int] = frozenset({12, 24, 36})
DEFAULT_COLUMNS = 12
#: What a new board is created with.
NEW_BOARD_COLUMNS = 24
MAX_WIDTH_PRESETS = ("1480", "full")
CUSTOM_WIDTH_RANGE = (320, 10000)


def _whole(value: Any) -> bool:
    return isinstance(value, int) and not isinstance(value, bool)


def columns_of(settings: dict[str, Any] | None) -> int:
    """The columns of a board, from its settings; twelve when it does not say."""
    value = (settings or {}).get("columns")
    return value if _whole(value) and value in ALLOWED_COLUMNS else DEFAULT_COLUMNS


def scale_size(size: tuple[int, int], columns: int) -> tuple[int, int]:
    """An adapter's size, declared in twelfths, on a board of ``columns``."""
    return size[0] * (columns // DEFAULT_COLUMNS), size[1]


def scale_layout(items: list[dict[str, Any]], from_cols: int, to_cols: int) -> list[dict[str, Any]]:
    """The same arrangement on a grid of ``to_cols``.

    Growing is exact. Shrinking rounds, keeps every card at least one column
    wide and pulls one that would hang over the right edge back in.
    """
    factor = to_cols / from_cols
    scaled: list[dict[str, Any]] = []
    for item in items:
        w = max(1, round(item["w"] * factor))
        x = min(max(0, round(item["x"] * factor)), to_cols - w)
        new = {**item, "x": x, "w": w}
        if "minW" in item:
            new["minW"] = max(1, round(item["minW"] * factor))
        scaled.append(new)
    return scaled


def normalise_settings(settings: dict[str, Any] | None) -> dict[str, Any]:
    """The settings with the three layout keys valid or gone; other keys as they came."""
    result = dict(settings or {})
    if "columns" in result and columns_of(result) != result["columns"]:
        del result["columns"]
    if "max_width" in result:
        width = result["max_width"]
        if not (width in MAX_WIDTH_PRESETS or (_whole(width) and CUSTOM_WIDTH_RANGE[0] <= width <= CUSTOM_WIDTH_RANGE[1])):
            del result["max_width"]
    if "fit_screen" in result:
        if isinstance(result["fit_screen"], bool | int):
            result["fit_screen"] = bool(result["fit_screen"])
        else:
            del result["fit_screen"]
    return result

File: frontend/src/lib/layout.ts
/**
 * The grid a board is arranged on: how many columns, how wide, how tall a row.
 *
 * All three are settings of the board. A board without them is drawn as
 * every board was before: twelve columns, 1480 px wide, rows of 68 px.
 * Adapters declare sizes in twelfths and are not touched; the scaling
 * happens here.
 */
export type Columns = 12 | 24 | 36
export const ROW_HEIGHT = 68
export const ROW_HEIGHT_COMPACT = 60
/** Under this a card cannot draw its title and a line; the board scrolls instead. */
export const ROW_HEIGHT_MIN = 48
/** Over this a three-row board on a 4K display turns into billboards. */
export const ROW_HEIGHT_MAX = 160
const DEFAULT_WIDTH = '1480px'
const CUSTOM_WIDTH = [320, 10000]

export function columnsOf(settings?: Record<string, unknown>): Columns {
  const value = settings?.columns
  return value === 24 || value === 36 ? value : 12
}

/** The CSS max-width of the board, or undefined when it may use it all. */
export function maxWidthOf(settings?: Record<string, unknown>): string | undefined {
  const value = settings?.max_width
  if (value === 'full') return undefined
  if (typeof value === 'number' && value >= CUSTOM_WIDTH[0] && value <= CUSTOM_WIDTH[1]) return `${value}px`
  return DEFAULT_WIDTH
}

/** The lowest row a card reaches; one for an empty page. */
export function rowsOf(items: { y: number; h: number }[]): number {
  return Math.max(1, ...items.map((item) => item.y + item.h))
}

/** The row height that makes `rows` rows fill `space` pixels, within what cards can bear. */
export function rowHeightFor(space: number, rows: number, gap: number): number {
  const raw = (space - gap * (rows - 1)) / Math.max(1, rows)
  return Math.floor(Math.min(ROW_HEIGHT_MAX, Math.max(ROW_HEIGHT_MIN, raw)))
}

/** How many board columns one twelfth is; never under one, so a phone's four columns cap rather than shrink. */
export function unitOf(columns: number): number {
  return Math.max(1, columns / 12)
}

/** An adapter's floor, declared in twelfths, on a board of `columns`. */
export function scaleFloor(size: [number, number], columns: number): [number, number] {
  return [size[0] * unitOf(columns), size[1]]
}

export type SizePreset = 'S' | 'M' | 'L' | 'XL'
export const SIZE_PRESETS: SizePreset[] = ['S', 'M', 'L', 'XL']

/**
 * Four sizes a card can be put to without dragging its corner: the smallest
 * the adapter says is usable, the size it was created with, and one and a
 * half and twice that. Width is capped at the grid; nothing goes below the
 * floor.
 */
export function presetSizes(widget: { default_size?: [number, number]; min_size?: [number, number] }, columns: number): Record<SizePreset, [number, number]> {
  const [floorW, floorH] = scaleFloor(widget.min_size ?? widget.default_size ?? [1, 1], columns)
  const [baseW, baseH] = scaleFloor(widget.default_size ?? widget.min_size ?? [3, 2], columns)
  const clamp = (w: number, h: number): [number, number] => [Math.max(1, Math.min(columns, Math.max(floorW, Math.round(w)))), Math.max(1, Math.max(floorH, Math.round(h)))]
  return {
    S: clamp(floorW, floorH),
    M: clamp(baseW, baseH),
    L: clamp(baseW * 1.5, baseH * 1.5),
    XL: clamp(baseW * 2, baseH * 2),
  }
}
